python/ray/rllib/utils/torch_policy_graph.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+from threading import Lock
+
+import torch
+import torch.nn.functional as F
+
+from ray.rllib.models.pytorch.misc import var_to_np
+from ray.rllib.utils.policy_graph import PolicyGraph
+
+
+class TorchPolicyGraph(PolicyGraph):
+    """Template for a PyTorch policy and loss to use with RLlib.
+
+    This is similar to TFPolicyGraph, but for PyTorch.
+
+    Attributes:
+        observation_space (gym.Space): observation space of the policy.
+        action_space (gym.Space): action space of the policy.
+        lock (Lock): Lock that must be held around PyTorch ops on this graph.
+    """
+
+    def __init__(
+            self, observation_space, action_space, model, loss, loss_inputs):
+        """Initialize the policy graph.
+
+        Arguments:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+            model (nn.Module): PyTorch policy module. Given observations as
+                input, this module must a list of outputs where the first item
+                are action logits, and the remainder can be any value.
+            loss (nn.Module): Loss defined as a PyTorch module. The inputs for
+                this module are defined by the `loss_inputs` param. This module
+                returns a single scalar loss.
+            loss_inputs (list): List of SampleBatch columns that will be
+                passed to the loss module's forward() function when computing
+                the loss. For example, ["obs", "action", "advantages"].
+
+        Attributes:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+        """
+        self.observation_space = observation_space
+        self.action_space = action_space
+        self.lock = Lock()
+        self._model = model
```

Comment:
this module doesn't do any accounting for where the model is (i.e., it could be on a gpu device or cpu device). This only matters if the model is on a gpu and the from_numpy() inputs below give cpu tensors (Which they do).

With the code as-is there should be a check here that verifies model device == device('cpu')

Reply:
or just docs saying input to the model is given as a cpu tensor

Reply:
I think we want to support both, and just migrate the data automatically to the right place like TensorFlow does.
So for model output, it sounds like .cpu().numpy() will do. Is there an equivalent for moving cpu tensors to the model device?

Reply:
well, you should probably make var_to_np == .cpu().detach().numpy() and then just use that since you always need those steps.

Inputs are tricky. The model doesn't have an associated device, can be spread out across several devices, or whatever. I can see two ways of doing this, neither too great:
1. Doc that your model should accept CPU tensors, then in the fwd pass users should move to whatever device they think is appropriate.
2. Assume the model is on a single device, which is the `.device` of the model's first parameter (then move all inputs `.to()` that)